### PR TITLE
Add tests for new credit card management flow

### PR DESCRIPTION
**Pull request by yashwin**
#### Changes proposed in this Pull Request

* This PR adds a few test cases to components, selectors, reducers, and actions to the code which was added as a part of new credit card management flow

`Note` Please run `yarn` or `yarn install` if you see the tests are failing due to missing modules

#### Testing instructions

* Run `git checkout update/add-tests-for-new-credit-card-management-flow && git pull`
* Since we have added tests under `client/jetpack-cloud/sections/partner-portal` run `yarn run test-client client/jetpack-cloud/sections/partner-portal` to run the tests. 
* Make sure all the tests are passing

Related to 1201734780767770-as-1201829766632184

### CODE REVIEW

**Comment by rcanepa** [comment]
@yashwin, I gave this PR a quick ran, and some tests failed. So here is the output I got:

```
❯ yarn run test-client client/jetpack-cloud/sections/partner-portal
 PASS  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-loading.jsx
 PASS  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/cvv-image.jsx
 PASS  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/payment-method-image.jsx
 FAIL  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-submit-button.jsx
  ● Test suite failed to run

    Cannot find module 'tus-js-client' from '../packages/wpcom.js/src/lib/tus-uploader.js'

    Require stack:
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/tus-uploader.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.media.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/index.js
      lib/wpcom-oauth-wrapper/index.js
      lib/wp/node.js
      state/partner-portal/partner/actions.ts
      state/partner-portal/partner/handlers.ts
      state/partner-portal/init.js
      state/partner-portal/credit-card-form/selectors.ts
      jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-submit-button.jsx

    > 1 | import * as tus from 'tus-js-client';
        | ^
      2 | import Media from './site.media';
      3 |
      4 | /**

      at Resolver.resolveModule (../node_modules/jest-resolve/build/resolver.js:324:11)
      at Object.<anonymous> (../packages/wpcom.js/src/lib/tus-uploader.js:1:1)

 PASS  client/jetpack-cloud/sections/partner-portal/add-stored-credit-card/test/add-stored-credit-card.jsx
 FAIL  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-fields.jsx
  ● Test suite failed to run

    Cannot find module 'tus-js-client' from '../packages/wpcom.js/src/lib/tus-uploader.js'

    Require stack:
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/tus-uploader.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.media.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/index.js
      lib/wpcom-oauth-wrapper/index.js
      lib/wp/node.js
      jetpack-cloud/sections/partner-portal/hooks.ts
      jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-fields.jsx

    > 1 | import * as tus from 'tus-js-client';
        | ^
      2 | import Media from './site.media';
      3 |
      4 | /**

      at Resolver.resolveModule (../node_modules/jest-resolve/build/resolver.js:324:11)
      at Object.<anonymous> (../packages/wpcom.js/src/lib/tus-uploader.js:1:1)

 FAIL  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/set-as-primary-payment-method.jsx
  ● <SetAsPrimaryPaymentMethod> › should render correctly, match the snapshot and fire change event correctly

    expect(received).toMatchSnapshot()

    Snapshot name: `<SetAsPrimaryPaymentMethod> should render correctly, match the snapshot and fire change event correctly 1`

    - Snapshot  - 1
    + Received  + 1

    @@ -46,11 +46,11 @@
                  <svg
                    aria-hidden="true"
                    class="components-checkbox-control__checked"
                    focusable="false"
                    height="24"
    -               role="presentation"
    +               role="img"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path

      26 | 		);
      27 |
    > 28 | 		expect( document.body ).toMatchSnapshot();
         | 		                        ^
      29 |
      30 | 		const input = document.body.getElementsByTagName( 'input' )[ 0 ];
      31 |

      at Object.<anonymous> (jetpack-cloud/sections/partner-portal/credit-card-fields/test/set-as-primary-payment-method.jsx:28:27)

 › 1 snapshot failed.
 FAIL  client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx
  ● Test suite failed to run

    Cannot find module 'tus-js-client' from '../packages/wpcom.js/src/lib/tus-uploader.js'

    Require stack:
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/tus-uploader.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.media.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/index.js
      lib/wpcom-oauth-wrapper/index.js
      lib/wp/node.js
      jetpack-cloud/sections/partner-portal/payment-methods/stored-payment-method-api.ts
      jetpack-cloud/sections/partner-portal/payment-methods/assignment-processor-functions.ts
      jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx
      jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx

    > 1 | import * as tus from 'tus-js-client';
        | ^
      2 | import Media from './site.media';
      3 |
      4 | /**

      at Resolver.resolveModule (../node_modules/jest-resolve/build/resolver.js:324:11)
      at Object.<anonymous> (../packages/wpcom.js/src/lib/tus-uploader.js:1:1)

Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or re-run jest with `-u` to update them.

Test Suites: 4 failed, 4 passed, 8 total
Tests:       1 failed, 4 passed, 5 total
Snapshots:   1 failed, 3 passed, 4 total
Time:        5.885 s
Ran all test suites matching /client\/jetpack-cloud\/sections\/partner-portal/i.
❯ yarn run test-client client/jetpack-cloud/sections/partner-portal/
 PASS  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-loading.jsx
 PASS  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/payment-method-image.jsx
 PASS  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/cvv-image.jsx
 PASS  client/jetpack-cloud/sections/partner-portal/add-stored-credit-card/test/add-stored-credit-card.jsx
 FAIL  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-submit-button.jsx
  ● Test suite failed to run

    Cannot find module 'tus-js-client' from '../packages/wpcom.js/src/lib/tus-uploader.js'

    Require stack:
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/tus-uploader.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.media.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/index.js
      lib/wpcom-oauth-wrapper/index.js
      lib/wp/node.js
      state/partner-portal/partner/actions.ts
      state/partner-portal/partner/handlers.ts
      state/partner-portal/init.js
      state/partner-portal/credit-card-form/selectors.ts
      jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-submit-button.jsx

    > 1 | import * as tus from 'tus-js-client';
        | ^
      2 | import Media from './site.media';
      3 |
      4 | /**

      at Resolver.resolveModule (../node_modules/jest-resolve/build/resolver.js:324:11)
      at Object.<anonymous> (../packages/wpcom.js/src/lib/tus-uploader.js:1:1)

 FAIL  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-fields.jsx
  ● Test suite failed to run

    Cannot find module 'tus-js-client' from '../packages/wpcom.js/src/lib/tus-uploader.js'

    Require stack:
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/tus-uploader.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.media.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/index.js
      lib/wpcom-oauth-wrapper/index.js
      lib/wp/node.js
      jetpack-cloud/sections/partner-portal/hooks.ts
      jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-fields.jsx

    > 1 | import * as tus from 'tus-js-client';
        | ^
      2 | import Media from './site.media';
      3 |
      4 | /**

      at Resolver.resolveModule (../node_modules/jest-resolve/build/resolver.js:324:11)
      at Object.<anonymous> (../packages/wpcom.js/src/lib/tus-uploader.js:1:1)

 FAIL  client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx
  ● Test suite failed to run

    Cannot find module 'tus-js-client' from '../packages/wpcom.js/src/lib/tus-uploader.js'

    Require stack:
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/tus-uploader.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.media.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/lib/site.js
      /Users/rcanepa/dev/automattic/wp-calypso/packages/wpcom.js/src/index.js
      lib/wpcom-oauth-wrapper/index.js
      lib/wp/node.js
      jetpack-cloud/sections/partner-portal/payment-methods/stored-payment-method-api.ts
      jetpack-cloud/sections/partner-portal/payment-methods/assignment-processor-functions.ts
      jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx
      jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx

    > 1 | import * as tus from 'tus-js-client';
        | ^
      2 | import Media from './site.media';
      3 |
      4 | /**

      at Resolver.resolveModule (../node_modules/jest-resolve/build/resolver.js:324:11)
      at Object.<anonymous> (../packages/wpcom.js/src/lib/tus-uploader.js:1:1)

 FAIL  client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/set-as-primary-payment-method.jsx
  ● <SetAsPrimaryPaymentMethod> › should render correctly, match the snapshot and fire change event correctly

    expect(received).toMatchSnapshot()

    Snapshot name: `<SetAsPrimaryPaymentMethod> should render correctly, match the snapshot and fire change event correctly 1`

    - Snapshot  - 1
    + Received  + 1

    @@ -46,11 +46,11 @@
                  <svg
                    aria-hidden="true"
                    class="components-checkbox-control__checked"
                    focusable="false"
                    height="24"
    -               role="presentation"
    +               role="img"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path

      26 | 		);
      27 |
    > 28 | 		expect( document.body ).toMatchSnapshot();
         | 		                        ^
      29 |
      30 | 		const input = document.body.getElementsByTagName( 'input' )[ 0 ];
      31 |

      at Object.<anonymous> (jetpack-cloud/sections/partner-portal/credit-card-fields/test/set-as-primary-payment-method.jsx:28:27)

 › 1 snapshot failed.
Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or re-run jest with `-u` to update them.

Test Suites: 4 failed, 4 passed, 8 total
Tests:       1 failed, 4 passed, 5 total
Snapshots:   1 failed, 3 passed, 4 total
Time:        4.616 s, estimated 5 s
```

Could a rebase fix these issues?

Just out of curiosity and because I don't know the answer. Do you know whether we want to use snapshots on Calypso? I have seen some of them, but not much, and I wonder whether that is because we have decided not to use them or because we haven't found them useful, or something else.

**Comment by vitozev** [comment]
> Just out of curiosity and because I don't know the answer. Do you know whether we want to use snapshots on Calypso?

IMPO, _snapshot tests_ can be helpful for user-critical flows at the point where we consider this flow a stable one. In this case, the payment methods flow is in the early development phase and may change quite a lot in the near future, and keeping the snapshots in sync feels like something we can avoid for now.

What do you think, @yashwin?

**Comment by yashwin** [comment]
> > Just out of curiosity and because I don't know the answer. Do you know whether we want to use snapshots on Calypso?
> 
> IMPO, _snapshot tests_ can be helpful for user-critical flows at the point where we consider this flow a stable one. In this case, the payment methods flow is in the early development phase and may change quite a lot in the near future, and keeping the snapshots in sync feels like something we can avoid for now.
> 
> What do you think, @yashwin?

Yes, makes sense to me. I'll go ahead and remove the snapshots from these tests. @vitozev @rcanepa

File: client/jetpack-cloud/sections/partner-portal/add-stored-credit-card/test/add-stored-credit-card.jsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+		const mockStore = configureStore();
+		const store = mockStore( initialState );
+
+		render(
+			<Provider store={ store }>
+				<AddStoredCreditCard />
+			</Provider>
+		);
+
+		const aTag = document.body.getElementsByClassName( 'add-stored-credit-card' )[ 0 ];
```
Comment (manzoorwanijk) [inline client/jetpack-cloud/sections/partner-portal/add-stored-credit-card/test/add-stored-credit-card.jsx]: Instead of using `document.body` context, you could directly use the container context returned by `render` to make it faster.

```suggestion
		const { container } = render(
			<Provider store={ store }>
				<AddStoredCreditCard />
			</Provider>
		);

		const [ aTag ] = container.getElementsByClassName( 'add-stored-credit-card' );
```

Comment (yashwin) [reply client/jetpack-cloud/sections/partner-portal/add-stored-credit-card/test/add-stored-credit-card.jsx]: @manzoorwanijk I have updated the tests to use containers like your suggestion. Please check - https://github.com/Automattic/wp-calypso/pull/62695/commits/0b95594fd94a95c8a328c63da8d3283810f831b3

**Comment by yashwin** [comment]
@vitozev, @rcanepa  I have removed the snapshot tests cases and updated few tests, please check here -https://github.com/Automattic/wp-calypso/pull/62695/commits/0b95594fd94a95c8a328c63da8d3283810f831b3

File: client/jetpack-cloud/sections/partner-portal/add-stored-credit-card/test/add-stored-credit-card.jsx (hunk 1 above)
Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/partner-portal/add-stored-credit-card/test/add-stored-credit-card.jsx]: Yeah, looks good now. Thank you for making those changes.

File: client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-fields.jsx [hunk 2]
```diff
@@ -0,0 +1,86 @@
+/**
+ * @jest-environment jsdom
+ */
+import { checkoutTheme } from '@automattic/composite-checkout';
+import { ThemeProvider } from '@emotion/react';
+import { Elements } from '@stripe/react-stripe-js';
+import { loadStripe } from '@stripe/stripe-js';
+import { useSelect, useDispatch } from '@wordpress/data';
+import '@testing-library/jest-dom/extend-expect';
+import { mount } from 'enzyme';
```
Comment (rcanepa) [inline client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-fields.jsx]: Out of curiosity. Why are we using `enzyme` instead of `@testing-library/react` like the rest of the tests?

Comment (yashwin) [reply client/jetpack-cloud/sections/partner-portal/credit-card-fields/test/credit-card-fields.jsx]: I think I missed moving this to `@testing-library/react`, have updated it now - https://github.com/Automattic/wp-calypso/pull/62695/commits/7fa129ef073af1b0fcfc3fdf7997f42746f6859b. Thanks for bringing this up.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		);
+
+		const [ aTag ] = container.getElementsByClassName( 'payment-method-add__back-button' );
+		const href = 'https://example.com/partner-portal/payment-methods/';
+
+		expect( aTag ).toHaveProperty( 'href', href );
+
+		expect( getByText( 'Credit card details' ) ).toBeInTheDocument();
+
+		await act( () => promise );
```
Comment (manzoorwanijk) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx:34]: Why is this needed here?

Comment (yashwin) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx:34]: The component `PaymentMethodAdd` has a `StripeSetupIntentIdProvider` which gets updated after the component is mounted. When testing, code that causes React state updates should be wrapped into an [act](https://reactjs.org/docs/test-utils.html#act). 

We get warnings like this otherwise:

Screenshot 2022-04-21 at 11 47 02 AM

This line is basically added so that it waits for all the promises to be resolved. Without this, the tests pass, but we get these warnings.

Comment (manzoorwanijk) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx:34]: Got it. Thank you.

File: client/state/partner-portal/credit-card-form/test/reducer.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+				[ key ]: {
+					errors: [ message ],
+					isTouched: true,
+					value: '',
+				},
+			},
+		} );
+	} );
+
+	test( 'should return an object by settting all the fields as touched', () => {
```
Comment (rcanepa) [inline client/state/partner-portal/credit-card-form/test/reducer.js:50]: There is a typo in `settting`.

**Review by rcanepa — APPROVED** [review]
LGTM.

These changes look good to me now. Thank you for adding these tests!

**Comment by yashwin** [comment]
> LGTM.
> 
> These changes look good to me now. Thank you for adding these tests!

@rcanepa - Thanks for the feedback and the review! 🙂
